### ISSUE TITLE
Multiple requests. Get ELF header and section information, PE section size reporting

### ISSUE THREAD

**Issue by cesenaLA** [davea42/libdwarf-code#206]
Hello! We recently updated one of our utilities to use `libdwarf 0.7.0` from an old version, the update went well and got everything working again, but we wanted to go one step further and stop using `libelf` ourselves. This update and the goal to remove `libelf` has prompt some requests to get more information that's already in the internal code of `libdwarf`. I've decided to summarize the requests on one issue, they are not very lengthy, but let me know if you prefer that I make multiple issues.

### Get Header information
We are interested in getting the processor type and endianness of the binary file that's being loaded. This information is already being loaded when reading the file into `libdwarf` internal structures but I haven't found a way to reach this information from the public API.

### Section information
Section information is something we are also interested in, and going through the documentation we found `dwarf_get_section_info_by_index` and `dwarf_get_section_count` to iterate through the sections available in the file. That worked great, but there are two things we still cannot get:
- Section header flags. We usually check a flag in elf section headers to verify if the section is allocatable (`SHF_ALLOC`). It would be nice if libdwarf could report if the section is allocatable or a way to get the section flags.
- Virtual size for PE sections. Iterating the sections of a PE file reports size = 0 for some sections. `libdwarf` does a `is_irrelevant_section` check for PE files https://github.com/davea42/libdwarf-code/blob/6c14db69a3819f43573144bba3d12e4dcdc006fb/src/lib/libdwarf/dwarf_peread.c#L369-L371
Which sets the `VirtualSize` and `SizeOfRawData` to 0 for the following sections:  https://github.com/davea42/libdwarf-code/blob/6c14db69a3819f43573144bba3d12e4dcdc006fb/src/lib/libdwarf/dwarf_peread.c#L342-L348
It removes them because they are irrelevant for dwarf information but it would be nice that at least `dwarf_get_section_info_by_index` reports the correct VirtualSize.

### Unify path separator
In cc110ae586ec920a3cd5bcab32f84a61f3a139ec libdwarf started to unify the paths reported to use '/' as its separator. The current function is `transfor_to_posix_slash`. And this works great, but the compilation directory (`DW_AT_comp_dir`) is still reported with the separator used during compilation. So in Windows some compilation directories end up with the '\\' slash. It looks like libdwarf actually gets the compilation directory directly from the dwarf information using a pointer, so normalizing the path separator would mean an additional allocation, so there's a tradeoff from this request, but maybe the consistency is worth it.

Those are all the requests, we currently get the information with `libelf`, but most of the information is already in `libdwarf` internal structures, so exposing that data could be an option.

Let me know what you think about the changes and if they are something we could get to see in libdwarf.

**Comment by davea42** [comment]
The object data libdwarf returns is useful for all supported object formats, whereas some Elf fields have no direct
equivalent in PE or MacOS objects.  So adding elf interfaces does not seem like an imperative.

Getting the correct VirualSize set properly for PE  sounds like a very good idea.  If you were to create
a Pull Request with that fixed I would approve it.

It will be a couple days before I have time, but I think the path-separator does deserve some thought.
Under Msys2 Windows the path-separator situation is muddled because the Windows
part of msys2 gets kind of mixed in with the posix part, depending on what shell commands or calls  one
is using.  the result being that running full regression tests on msys2 has more fails
than I am comfortable with (though most tests pass).  

I don't quite understand if you are suggesting that libdwarf should modify all paths everywhere to '\'  
in strings reported from libdwarf?  That may not play well
at all with msys2!
Or are you suggesting something less? Or ?

**Comment by cesenaLA** [comment]
> The object data libdwarf returns is useful for all supported object formats, whereas some Elf fields have no direct
> equivalent in PE or MacOS objects. So adding elf interfaces does not seem like an imperative.

What if we could settle on returning the raw value of the section flags and the raw cpu type value. ELF, PE and Mach-o have those flags available, they are just interpreted differently. The three formats have a Dwarf_Unsigned flag field in their section header struct:
- `generic_shdr`
- `generic_macho_section`
- `dwarf_pe_generic_image_section_header`

`libdwarf` could have a function to retrieve the raw value of those flags and have the user actually decode what those flags mean. The user can get the file type with one of the `dwarf_object_detector_*` functions with the `dw_ftype` parameter. And now that I look at those function, I can get the endianness from them :), so that request can be ignored now.

The flags are read here:  https://github.com/davea42/libdwarf-code/blob/6c14db69a3819f43573144bba3d12e4dcdc006fb/src/lib/libdwarf/dwarf_init_finish.c#L1683-L1692

The field is set for elf https://github.com/davea42/libdwarf-code/blob/6c14db69a3819f43573144bba3d12e4dcdc006fb/src/lib/libdwarf/dwarf_elfread.c#L186-L196

For PE files it can be set to the [Characteristics](https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#section-flags) field.  https://github.com/davea42/libdwarf-code/blob/6c14db69a3819f43573144bba3d12e4dcdc006fb/src/lib/libdwarf/dwarf_peread.c#L239-L257
The same could be done for the cpu type, get the raw cpu flag from the file header and have the user figure out what the value means.

And for Mach-o it would be setting it here to `generic_macho_section::flags` https://github.com/davea42/libdwarf-code/blob/6c14db69a3819f43573144bba3d12e4dcdc006fb/src/lib/libdwarf/dwarf_machoread.c#L192-L207

> Getting the correct VirualSize set properly for PE sounds like a very good idea. If you were to create
> a Pull Request with that fixed I would approve it.

Great! I'll work on that (it might also take me some time to get the PR open).

> I don't quite understand if you are suggesting that libdwarf should modify all paths everywhere to '`\\

Yeah that was the basic suggestion, and it's more to normalize the separator used by libdwarf now that all other paths seems to be converted to the posix slash. But if changing it for the compilation directory breaks something, I don't think is worth it. I'm not very familiar with msys2 so I will have to check, thanks for letting me know!

**Comment by davea42** [comment]
In dwarf_peread.c when requesting section information with pe_get_section_info(), we do
`return_section->as_size = sp->VirtualSize;`
to return such (line 245).

Is that not correct?

**Comment by davea42** [comment]
In an Elf file, just read the first 16 bytes of the file into an array of  'unsigned char'
and you have the basic data you are interested in, Right? Easy Peasy.

**Comment by cesenaLA** [comment]
> In dwarf_peread.c when requesting section information with pe_get_section_info(), we do
> return_section->as_size = sp->VirtualSize;
> to return such (line 245).

Yes, but sp->VirtualSize is set to 0 for sections  ".text",  ".bss",  ".data", and ".rdata". Just after reading them, if the section name matches one of those the function `is_irrelevant_section` sets its VirtualSize to 0. In a comment is mentioned that is to prevent an allocation for those sections, maybe we can add a "is_relevant" flag to prevent the allocation instead of setting their sizes to 0.

**Comment by davea42** [comment]
The change breaks nothing in regression tests.  If you apply that patch/changes to your copy does
it now do what you want?

**Comment by cesenaLA** [comment]
Confirmed. It works with the patch now, thanks!

**Comment by davea42** [comment]
Committed and pushed to github. Enjoy!

**Comment by davea42** [comment]
Note that dwarf_object_detector_path_b() will give you some of what you want:
endianness, and offsetsize (object level offsetsize and endianness, not DWARF level).
And it informs which type of object as well: PE, MacOS, or Elf.

The processor type, however, is not currently available through the API.
Hmm.

**Comment by davea42** [comment]
One could imagine a new API function returning
```
type of object file (a value defined by libdwarf)
endianness
object_format pointer size
processor type
processor subtype (relevant for MacOS)
processor flags
```
This would be just a little new code...not much.

What do you think?



```
On basic object data Nov 29,2023:
In an Elf header, 

e_machine identifies  the machine involved.
For example, EM_386 for older Intel.
See "elf.h" for the actual values.

e_ident[EI_DATA] has either
  ELFDATA2LSB or
  ELFDAAT2MSB
indicating endianness.

e_flags has processor-related flags of the
form EF_<varions>..


In a Mach-O header,

If it is a universal-binary see struct fat_arch
or struct fat_arch_64:
cputupe and cputype have machine identification values. 
(32bit signed int)

For a specfic actual object (possibly by itself
or possibly one such in a universal-binary)
struct mach_header or mach_header64:
cputupe and cputype have machine identification values. 
(32bit signed int)

Endianness is derived from reading the magic-number
field in mach_header or mach_header_64.
which will be MH_MAGIC    0xfeedface (big endian) or
MH_CIGAM    0xcefaedfe (little-endian).

The flags field is 4 bytes long and its possible
values are documented in the flags list MH_NOUNDEFS
being one example.


n a PE header set:

The first two bytes of the object, a dos_header,
are either  0x 5a 0x4d (little endian)
or  0x4d 0x5a (big endian).

The next struct is IMAGE_FILE_HEADER
and its first two bytes are a Machine designator.
For examplle IMAGE_FILE_MACHINE_I386
Its last two bytes are a Characteristics flag set.
For example IMAGE_FILE_32BIT_MACHINE  
See: 
learn.microsoft.com/en-us/windows/win32/debug/pe-format#machine-types.


```

**Comment by cesenaLA** [comment]
Yeah, I function like that would certainly help. In our case we don't need the additional processor flags at the moment (`e_flags`, PE characteristics, etc) but its nice to have. The flags that we were looking to get are the section flags (`sh_flags` for elf for example, PE and Mach-o have similar section flags).

**Comment by davea42** [comment]
In PE it is section Characteristics (it appears) and that would be returned from
a get_section_info (the existing function) as as_flags.
Though  Dwarf_Obj_Access_Section_a struct. is a public struct and Dwarf_Obj_Access_Methods_a_s
is a public struct Dwarf_Obj_Access_Section_a_s is public no API
functions allow access directly.

All those need to be public so that dwarf_object_init_b() can let people read their own object formats via
code they write themselves.

Hmm.  I wonder if there is a way forward that was not noticed earlier.

**Comment by davea42** [comment]
I have created a new API entry:
```
DW_API int dwarf_get_object_overview(Dwarf_Debug dw_dbg,
+    Dwarf_Small    *dw_ftype,
+    Dwarf_Small    *dw_obj_pointersize,
+    Dwarf_Bool     *dw_obj_is_big_endian,
+    Dwarf_Unsigned *dw_obj_machine,
+    Dwarf_Unsigned *dw_obj_flags,
+    Dwarf_Small    *dw_path_source,
+    Dwarf_Unsigned *dw_ub_offset,
+    Dwarf_Unsigned *dw_ub_count,
+    Dwarf_Unsigned *dw_ub_index,
+    Dwarf_Unsigned *dw_comdat_groupnumber);
```


/tmp/dwarfdump -v --print-obj-overview /tmp/dwarfdump

```
Selected Object Overview fields
  Filetype              : 1       DW_FTYPE_ELF
  Pointersize           : 8
  endian                : little endian
  machine/architecture  : 62 0x3e
  machine flags         : 0x0
  path source           : 1       DW_PATHSOURCE_basic
  Universal Binary 
    Object Offset in UB : Ox0
    UB object count     : 0
    UB index            : 0
  Comdat group number   : 1

```
All these fields exist for all object formats, though the machine and flags
content varies by object format (and obviously by the machine the object was built for).

So, my question.  If this gets published will it meet your needs?  Don't forget
dwarf_get_section_by_name_a() which has new fields  (not showing up publically yet, 

In sum, for any object file, it is now possible to take (for example) a GOFF= global  or a local .debug_info
offset and combine with a section offset, (and if a MacOS universal-binary combine with an inner-executable
offset ) to get a number which is the physical file offset of  that DIE or whatever.

But I do not know if that actually matters to you...

**Comment by cesenaLA** [comment]
> So, my question. If this gets published will it meet your needs? 
Yes! that's one of the missing part that we where looking for.

> Don't forget dwarf_get_section_by_name_a() which has new fields (not showing up publically yet,
What would be the new fields, if they include the section flags field, yes, it will meet our needs.

Thanks for all the support.

**Comment by cesenaLA** [comment]
I tested the new API functions and its working great!

The only remaining topic is the '/' slash change to compiler directory, but that is a nitpick to have it to be the same across `libdwarf`, if that breaks the code is not worth it. Let me know if you want to keep the issue open for that. If not I think we can close it as completed.

Thanks again for the quick support!
